File: src/test/java/com/fwdekker/randomness/DataInsertActionTest.java
package com.fwdekker.randomness;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.openapi.editor.Editor;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;


/**
 * Unit tests for {@link DataInsertAction}.
 */
final class DataInsertActionTest {
    /**
     * The recognizable string that is inserted by the insertion action.
     */
    private static final String RANDOM_STRING = "random_string";

    private DataInsertAction dataInsertAction;


    @BeforeEach
    void beforeEach() {
        dataInsertAction = new SimpleInsertAction();
    }


    /**
     * Tests that no further actions are taken when the editor is {@code null}.
     */
    @Test
    void testActionPerformedNullEditor() {
        final AnActionEvent event = mock(AnActionEvent.class);
        when(event.getData(CommonDataKeys.EDITOR)).thenReturn(null);

        dataInsertAction.actionPerformed(event);

        verify(event, times(1)).getData(CommonDataKeys.EDITOR);
        verifyNoMoreInteractions(event);
    }

    /**
     * Tests that the action's presentation is disabled when the editor is null.
     */
    @Test
    void testUpdateDisabled() {
        final AnActionEvent event = mock(AnActionEvent.class);
        final Presentation presentation = spy(Presentation.class);
        when(event.getData(CommonDataKeys.EDITOR)).thenReturn(null);
        when(event.getPresentation()).thenReturn(presentation);

        dataInsertAction.update(event);

        assertThat(presentation.isEnabled()).isFalse();
    }

    /**
     * Tests that the action's presentation is enabled when the editor is not null.
     */
    @Test
    void testUpdateEnabled() {
        final AnActionEvent event = mock(AnActionEvent.class);
        final Presentation presentation = spy(Presentation.class);
        when(event.getData(CommonDataKeys.EDITOR)).thenReturn(mock(Editor.class));
        when(event.getPresentation()).thenReturn(presentation);

        dataInsertAction.update(event);

        assertThat(presentation.isEnabled()).isTrue();
    }


    /**
     * Simple implementation of {@code DataInsertAction}.
     */
    private static class SimpleInsertAction extends DataInsertAction {
        @Override
        protected String getName() {
            return "Insert Simple";
        }

        @Override
        protected String generateString() {
            return RANDOM_STRING;
        }
    }
}

File: src/test/kotlin/com/fwdekker/randomness/DataInsertActionTest.kt
package com.fwdekker.randomness

import com.fwdekker.randomness.DataInsertAction.Companion.GENERATOR_TIMEOUT
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.actionSystem.Presentation
import com.intellij.openapi.editor.Document
import com.intellij.testFramework.fixtures.BasePlatformTestCase
import icons.RandomnessIcons
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe
import kotlin.math.roundToLong
import kotlin.random.Random


/**
 * Unit tests for [DataInsertAction].
 *
 * @see DataInsertActionIntegrationTest
 */
class DataInsertActionTest : BasePlatformTestCase() {
    private lateinit var dataInsertAction: DataInsertAction
    private lateinit var document: Document


    override fun setUp() {
        super.setUp()

        val file = myFixture.copyFileToProject("emptyFile.txt")
        myFixture.openFileInEditor(file)

        document = myFixture.editor.document

        dataInsertAction = DummyInsertAction { "random_value" }
    }

    override fun getTestDataPath() = javaClass.classLoader.getResource("integration-project/")?.path


    fun testActionPerformedNoFurtherActionsWhenEditorIsNull() {
        val event = AnActionEvent.createFromDataContext("", null) {
            if (it == CommonDataKeys.PROJECT.name) myFixture.project
            else null
        }

        dataInsertAction.actionPerformed(event)

        assertThat(document.text).isEqualTo("")
    }

    fun testActionPerformedNoFurtherActionsWhenProjectIsNull() {
        val event = AnActionEvent.createFromDataContext("", null) {
            if (it == CommonDataKeys.EDITOR.name) myFixture.editor
            else null
        }

        dataInsertAction.actionPerformed(event)

        assertThat(document.text).isEqualTo("")
    }

    fun testUpdateDisablePresentationWhenEditorIsNull() {
        val presentation = Presentation()
        val event = AnActionEvent.createFromDataContext("", presentation) { null }

        dataInsertAction.update(event)

        assertThat(presentation.isEnabled).isFalse()
    }

    fun testUpdateEnablePresentationWhenEditorIsNotNull() {
        val presentation = Presentation()
        val event = AnActionEvent.createFromDataContext("", presentation) { myFixture.editor }

        dataInsertAction.update(event)

        assertThat(presentation.isEnabled).isTrue()
    }
}


/**
 * Unit tests for timeliness of generators in [DataInsertAction].
 */
class DataInsertActionTimeoutTest : Spek({
    /**
     * Inserts a dummy value after a delay that is longer than [GENERATOR_TIMEOUT].
     *
     * @property dummySupplier generates dummy values to insert
     */
    class SlowDummyInsertAction(private val dummySupplier: (Random) -> String) :
        DataInsertAction(RandomnessIcons.Data.Base) {
        override val name = "Slow Random Dummy"

        override fun generateStrings(count: Int): List<String> {
            Thread.sleep((1.1 * GENERATOR_TIMEOUT).roundToLong())
            return List(count) { dummySupplier(random) }
        }
    }


    describe("timely generators") {
        lateinit var action: SlowDummyInsertAction


        beforeEachTest {
            action = SlowDummyInsertAction { "test string" }
        }


        it("generates a string normally") {
            assertThat(action.generateString()).isEqualTo("test string")
        }

        it("generates multiple strings normally") {
            assertThat(action.generateStrings(2)).isEqualTo(listOf("test string", "test string"))
        }

        it("fails to generate a single string in time") {
            assertThatThrownBy { action.generateStringTimely() }
                .isInstanceOf(DataGenerationException::class.java)
                .hasMessage("Timed out while generating data.")
        }

        it("fails to generate multiple strings in time") {
            assertThatThrownBy { action.generateStringsTimely(2) }
                .isInstanceOf(DataGenerationException::class.java)
                .hasMessage("Timed out while generating data.")
        }
    }
})
